tests/integration/helpers.py

```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Helpers for the slurmdbd integration tests."""
+
+import contextlib
+import logging
+import pathlib
+import shlex
+import subprocess
+import tempfile
+from io import StringIO
+from typing import Dict
+
+import paramiko
+import tenacity
+from pytest_operator.plugin import OpsTest
+from urllib import request
+
+logger = logging.getLogger(__name__)
+
+ETCD = "etcd-v3.5.0-linux-amd64.tar.gz"
+ETCD_URL = f"https://github.com/etcd-io/etcd/releases/download/v3.5.0/{ETCD}"
+VERSION = "version"
+VERSION_NUM = subprocess.run(
+    shlex.split("git describe --always"), stdout=subprocess.PIPE, text=True
+).stdout.strip("\n")
+
+
+def get_slurmctld_res() -> Dict[str, pathlib.Path]:
+    """Get slurmctld resources needed for charm deployment."""
+    if not (etcd := pathlib.Path(ETCD)).exists():
+        logger.info(f"Getting resource {ETCD} from {ETCD_URL}...")
+        request.urlretrieve(ETCD_URL, etcd)
+
+    return {"etcd": etcd}
+
+
+def get_slurmdbd_res() -> None:
+    """Get slurmdbd charm resources needed for deployment."""
+    if not (version := pathlib.Path(VERSION)).exists():
+        logger.info(f"Setting resource {VERSION} to value {VERSION_NUM}...")
+        version.write_text(VERSION_NUM)
+
+
+@contextlib.asynccontextmanager
+async def unit_connection(
+    ops_test: OpsTest, application: str, target_unit: str
+) -> paramiko.SSHClient:
```

Comment:
The ContextManager to do the SSH is a fairly clean solution, however I think it's unnecessary and adds a bit of complexity regarding the handling of ssh known hosts, key file management, etc. Using the [unit.ssh](https://github.com/juju/python-libjuju/blob/master/juju/unit.py#L311) method allows you to use the keys juju is already using for the environment. An alternative to using the `unit.ssh` is to use the [unit.run](https://github.com/juju/python-libjuju/blob/master/juju/unit.py#L189) method, which returns an `ActionResult` rather than a `str`.

Now, with the `unit.run` - you'll want to understand that this will run in juju context- which means that you'll have some environment variables that are set by Juju regarding which unit you are accessing and that you'll also have to wait to acquire any machine locks necessary to execute (Juju allows one unit per machine to hold the machine lock and execute certain types of commands). As such, you may see some delay when executing the `unit.run` but when it does execute you know that other units on the machine will not be running a hook which can alter the state. I don't believe this level of isolation is necessary for the test, but I wanted to take a moment to spell it out.

I'll also note that there are subtle differences between your code and the unit.ssh code. For example, while not documented, the unit.ssh method will raise a `JujuError` when the [return code of the subprocess is not 0](https://github.com/juju/python-libjuju/blob/b90f16ba0d4670f159aaccda6dfab8135826abe4/juju/machine.py#L162) and capture the command and the stderr in the exception message.

Reply:
Provided @dvdgomez feedback on slurmctld-operator, it looks like the asynchronous context manager is going to my personal gist pile :sweat_smile:

I will experiment with using `unit.ssh` over the context manager. I think the context manager is a "clever" solution, but I will concede that it is complex up front with all it's bells and whistles for addressing latency with Juju.

Reply:
Resolved in https://github.com/canonical/slurmdbd-operator/pull/4/commits/05ac3889ef35a7fc621427d571247016b43f8dca. I saved the asynchronous context manager as a public gist on GitHub here so that we can revisit it later should we need it: https://gist.github.com/NucciTheBoss/7f2fe9a000403152416c5272f9f0bed9